### ISSUE TITLE
Provide ui test infrastructure for testing multi quickfixes [2/3]

### ISSUE THREAD

**Issue by drkstr101** [eclipse/xtext-xtend#1061]
This is a follow up on: https://github.com/eclipse/xtext-eclipse/issues/1447

As the next step will take place in this repository, I am opening up an issue to create the `EqualsWithNullMultiQuickfixTest` and `RemoveUnnecessaryModifiersMultiQuickfixTest` test cases described in the referenced issue above.

**Comment by miklossy** [comment]
@drkstr101 Are you still working on this issue? Please let us know if you cannot manage, in that case I can take care of this.

**Comment by drkstr101** [comment]
@miklossy I tried working on this last week but I kept getting install errors when attempting to setup 2020-06 contributors bundle. I was thinking maybe it was just overloaded due to the recent release, so I will try again now and see if the weekend makes any difference. I will post back once I'm up and running again. I have should have this wrapped up soon after that.

**Comment by cdietrich** [comment]
its still broken right now because https://github.com/eclipse/xtext-eclipse/pull/1479 does not get reviewed

**Comment by drkstr101** [comment]
Will it cause any problems if I just continue to build off 2020-03?

I still get:

    No repository found at https://download.eclipse.org/eclipse/updates/4.16-I-builds.

On 2020-06 Install.

**Comment by cdietrich** [comment]
Yes. If you switch to my branch for Xtext-eclipse it should work

**Comment by drkstr101** [comment]
Yeah I still got the same error with 2020-03, so let me give that I try.

Thanks for the help!

**Comment by drkstr101** [comment]
@miklossy My apologies for being AFK about a week. I was able to take a bit of personal time though, and am all refreshed and ready to tackle this again. I am now up and running on `2020-06` without issue, and have a totally free schedule this week. My current mission is to finish out this task or bust!

**Comment by drkstr101** [comment]
When I attempt to run `xtend.ide.tests.slow (xtend).launch`, I am getting:

    The input element of the launch configuration does not exist

I was unable to locate any additional context. The other launchers work as expected.

My understanding is that my tests should go in `org.eclipse.xtend.ide.tests/longrunning/src`, and so I would assume this is the launcher I'm supposed to use to kick off the tests. I was originally putting my tests next to `SpellingQuickfixTest`, but I think this was incorrect based on the package name shown in the example stub.

I will see if I can kick these off with Maven or Gradle, but if anyone knows what the issue is off the top of their head please speak up. Or if this is unexpected behavior then I can open up another issue and get to work on it. Any additional insight would be greatly appreciated.

Cheers!

**Comment by cdietrich** [comment]
This is a known problem

https://github.com/eclipse/xtext/issues/1493

**Comment by drkstr101** [comment]
My apologies! I scanned the open issues before posting, but I guess I missed it. I will see if I can make any progress on that ticket.

Cheers!

**Comment by drkstr101** [comment]
Well, I was able to get my first test up and running but unfortunately, the results are not what I was expecting.

```groovy
class Foo {
	def foo(Object x) {
		return if(x == null) 0 else 1
	}
}
```

I would expect 1 quickfix proposal but am getting 0.  Since this works fine in my IDE I would assume the error is in my test case. Oddly, I was able to copy in the [testSpellingQuickfixInString](https://github.com/drkstr101/xtext-xtend/pull/1/files#diff-712d3357f55d9e3fa71f4f285e4aeb6fR59) test case into my own class, and this one does pass as expected.

I will keep plugging away at it. I just wanted to give a quick update on my progress. I've opened a [PR](https://github.com/drkstr101/xtext-xtend/pull/1) in my fork if you wish to see this for yourself.

**Comment by miklossy** [comment]
Hello Aaron,

In order to make the `EqualsWithNullMultiQuickfixTest` test case work, you have to ensure the following things:
- create a plugin project before the test execution and place the dsl file into a package within its src folder
- after setting the xtextEditor document but before invoking the computeQuickAssistProposals method, wait for the Reconciler to finish
- calculate the right offset and pass it to the computeQuickAssistProposals method

These can be achieved by the following modifications in your `EqualsWithNullMultiQuickfixTest` class:

```xtend
...
import org.eclipse.xtext.ui.refactoring.ui.SyncUtil
import static extension org.eclipse.xtend.ide.tests.WorkbenchTestHelper.createPluginProject
...

class EqualsWithNullMultiQuickfixTest extends AbstractMultiQuickfixTest {
	
	@Inject extension SyncUtil
	...
	static final String SINGLE_EQUALS_NULL_IN_EXPRESSION = '''
		package foo
		class Foo {
			...
		}
	...
	override dslFile(String projectName, String fileName, String fileExtension, CharSequence content) {
		projectName.createPluginProject
		super.dslFile(projectName, "src/foo/" + fileName, fileExtension, content)
	}
	...
	@Test def testSingleEqualsNullQuickfixInExpression() {
		xtextEditor.getDocument().set(SINGLE_EQUALS_NULL_IN_EXPRESSION)
		xtextEditor.waitForReconciler
		val offset = SINGLE_EQUALS_NULL_IN_EXPRESSION.indexOf("==") + 1
		var ICompletionProposal[] quickAssistProposals = computeQuickAssistProposals(offset)
		...
	}
	...
}
```

Hope that helps,
Tamás

**Comment by drkstr101** [comment]
Thank you so much Tamás, supper helpful as always! =)

**Comment by drkstr101** [comment]
[Still no joy](https://github.com/drkstr101/xtext-xtend/blob/b0e8e0413875a3a45846f99cbaef9d703752873b/org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/EqualsWithNullMultiQuickfixTest.xtend), but I feel that I am getting closer. At least I understand better what needs to happen and what the issue is.

I believe there is still some issue in the initial project setup. When running other tests I can see the content flashing in the editor for a brief moment, but this doesn't happen when running my own as a single unit. I will keep at it, but please do let me know if you notice an obvious mistake.

Cheers!

**Comment by miklossy** [comment]
There is a small mistake in the `SINGLE_EQUALS_NULL_IN_EXPRESSION`. 
Instead of 

```xtend
package foo {
	class Foo {
		...
	}
}
```

you have to write 


```xtend
package foo
class Foo {
	...
}
```

**Comment by cdietrich** [comment]
@miklossy why is this?

**Comment by miklossy** [comment]
The test case will not succeed if the [initial DSL text](https://github.com/drkstr101/xtext-xtend/blob/b0e8e0413875a3a45846f99cbaef9d703752873b/org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/EqualsWithNullMultiQuickfixTest.xtend#L39-L46) is not a proper Xtend content.

**Comment by cdietrich** [comment]
got it. was confused by improper indentation on your example

**Comment by drkstr101** [comment]
Oh wow, good eye. I of course know how to declare a package haha. I must have just kept glancing over my "to stupidly easy to be incorrect" example as unimportant.

Maybe next time I should sleep on it before wasting anybody's time 😏

**Comment by drkstr101** [comment]
[Hello, AbstractMultiQuickfixTest!](https://github.com/drkstr101/xtext-xtend/blob/d8ae53085c294c5d879b969e91946923caed0251/org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/EqualsWithNullMultiQuickfixTest.xtend)

I will write out some more cases for `EqualsWithNullMultiQuickfixTest` and `RemoveUnnecessaryModifiersMultiQuickfixTest`, then give my tentative plan on cleaning up `AbstractMultiQuickfixTest` to get your feedback before I start on that task.

Cheers, and thanks again for the help getting started!

**Comment by miklossy** [comment]
@drkstr101 Would be nice if some of these test cases could make use of the API provided by https://github.com/eclipse/xtext-eclipse/pull/1486

@cdietrich Can we merge that PR as soon as possible? Later on, we can refactor it any time if necessary.

**Comment by drkstr101** [comment]
@miklossy Yes indeed, I will get on that!

**Comment by drkstr101** [comment]
Ah yes, this new functionality should address my `// TODO Assert that quickfix picks up all markers` comment. I will hold off for a short while as things are wrapped up over there. In the meantime I will be stubbing out my test cases.

**Comment by miklossy** [comment]
That PR has just been merged, so that you can continue.

**Comment by drkstr101** [comment]
@miklossy I screwed up my workstation yesterday upgrading to Ubuntu20. I was finally able to get to a command line today, so hopefully I will be back up and running today. I will post back once I am.

**Comment by drkstr101** [comment]
@miklossy I screwed up my workstation yesterday upgrading to Ubuntu20. I was finally able to get to a command line today, so hopefully I will be back up and running today. I will post back once I am.

**Comment by drkstr101** [comment]
And I am back in business! My goal is to have some test cases ready for review by the end of the day. Cheers!

**Comment by drkstr101** [comment]
@miklossy 

Now I am getting `java.lang.AssertionError: expected:<1> but was:<0>` from [this line](https://github.com/eclipse/xtext-eclipse/blob/013cebe66d279b43ab0c79392226a43cd05edcb6/org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractMultiQuickfixTest.java#L145)

Am I missing something in my test case maybe?

```groovy
val initialText = '''
    package foo
    class Foo {
        def m(Object a, Object b, Object c) {
            if(a == null || b != null || c === null) 0 else 1
        }
    }
'''
val initialTextWithMarkers = '''
    package foo
    class Foo {
        def m(Object a, Object b, Object c) {
            if(a <0<==>0> null || b <1<!=>1> null || c === null) 0 else 1
        }
    }
    ---------------------------------------------------------------------
    0: message=The operator '==' should be replaced by '===' when null is one of the arguments.
    1: message=The operator '!=' should be replaced by '!==' when null is one of the arguments.
'''
val resultTextWithMarkers = '''
    package foo
    class Foo {
        def m(Object a, Object b, Object c) {
            if(a === null || b !== null || c === null) 0 else 1
        }
    }
    ---------------------------------------------------------
    (no markers found)
'''

// TODO java.lang.AssertionError: expected:<1> but was:<0>
testMultiQuickfix(initialText, initialTextWithMarkers, resultTextWithMarkers)
```

**Comment by miklossy** [comment]
@drkstr101 : Do you have the entire class uploaded somewhere?

By the way, I prefer to write the test cases in a more compact form, such as:
```
	@Test def test() {
		'''
		    package foo
		    class Foo {
		        def m(Object a, Object b, Object c) {
		            if(a == null || b != null || c === null) 0 else 1
		        }
		    }
		'''.testMultiQuickfix('''
		    package foo
		    class Foo {
		        def m(Object a, Object b, Object c) {
		            if(a <0<==>0> null || b <1<!=>1> null || c === null) 0 else 1
		        }
		    }
		    ---------------------------------------------------------------------
		    0: message=The operator '==' should be replaced by '===' when null is one of the arguments.
		    1: message=The operator '!=' should be replaced by '!==' when null is one of the arguments.
		''', '''
		    package foo
		    class Foo {
		        def m(Object a, Object b, Object c) {
		            if(a === null || b !== null || c === null) 0 else 1
		        }
		    }
		    ---------------------------------------------------------
		    (no markers found)
		''')
	}
```

**Comment by drkstr101** [comment]
Yes I have it in my [drkstr101/task_161](https://github.com/drkstr101/xtext-xtend/blob/task_1061/org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/EqualsWithNullMultiQuickfixTest.xtend) branch.

I will update the test case to match your code style. I was planning on having one test case using the xtend api, and the other using the java style api. Please let me know if you would like them all to use `testMultiQuickfix` instead.

**Comment by miklossy** [comment]
> I was planning on having one test case using the xtend api, and the other using the java style api.

Good idea!

**Comment by miklossy** [comment]
Seems there is a problem if you open the Xtext editor in the [test setup method](https://github.com/drkstr101/xtext-xtend/blob/task_1061/org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/EqualsWithNullMultiQuickfixTest.xtend#L62).

The following version works for me:
```
/** 
 * Copyright (c) 2012, 2020 itemis AG (http://www.itemis.eu) and others.
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 * SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.xtend.ide.tests.quickfix

import java.util.Arrays
import org.eclipse.xtend.ide.tests.XtendIDEInjectorProvider
import org.eclipse.xtext.testing.InjectWith
import org.eclipse.xtext.testing.XtextRunner
import org.eclipse.xtext.ui.editor.quickfix.QuickAssistCompletionProposal
import org.eclipse.xtext.ui.testing.AbstractMultiQuickfixTest
import org.junit.Test
import org.junit.runner.RunWith

import static extension org.eclipse.xtend.ide.tests.WorkbenchTestHelper.createPluginProject

/** 
 * @author Aaron R Miller - Initial contribution and API
 */
@RunWith(XtextRunner)
@InjectWith(XtendIDEInjectorProvider)
class EqualsWithNullMultiQuickfixTest extends AbstractMultiQuickfixTest {

	override void setUp() {
		super.setUp
		projectName.createPluginProject
	}

	@Test def testEqualsNullQuickfixInExpression() {
		'''
			package foo
			class Foo {
				def m(Object a, Object b, Object c) {
					if(a == null || b != null || c === null) 0 else 1
				}
			}
		'''.testMultiQuickfix('''
			package foo
			class Foo {
				def m(Object a, Object b, Object c) {
					if(a <0<==>0> null || b <1<!=>1> null || c === null) 0 else 1
				}
			}
			---------------------------------------------------------------------
			0: message=The operator '==' should be replaced by '===' when null is one of the arguments.
			1: message=The operator '!=' should be replaced by '!==' when null is one of the arguments.
		''','''
			package foo
			class Foo {
				def m(Object a, Object b, Object c) {
					if(a === null || b !== null || c === null) 0 else 1
				}
			}
			-----------------------------------------------------------
			(no markers found)
		''')
	}

	@Test def testEqualsNullQuickfixInSwitch() {
		val content = '''
			package foo
			class Foo {
				def m(Object a, Object b, Object c) {
					switch true {
						case a == null: 0
						case b != null: 0
						case c === null: 0
						default: 1
					}
				}
			}
		'''
		val xtextEditor = openEditor(dslFile(content))	

		val offset = content.indexOf("==") + 1
		val proposals = Arrays.asList(xtextEditor.computeQuickAssistProposals(offset))

		assertEquals(1, proposals.size)
		assertEquals(1, proposals.filter[it instanceof QuickAssistCompletionProposal].size)
		assertEquals("Replace '==' with '===' and '!=' with '!=='", proposals.get(0).displayString)
	}

	override getFileName() {
		"Foo"
	}

	override dslFile(CharSequence content) {
		super.dslFile(projectName, "src/foo/" + fileName, fileExtension, content)
	}

}
```

**Hint:** If you write the test code like that, there is no need to call the waitForReconclier explicitely.

**Comment by drkstr101** [comment]
@miklossy Thanks for the assist on this one! I will have a patch up here shortly. I've also added some cases to `RemoveUnnecessaryModifiersMultiQuickfixTest`, but they need to be cleaned up.

**Comment by drkstr101** [comment]
@miklossy 

[EqualsWithNullMultiQuickfixTest](https://github.com/drkstr101/xtext-xtend/blob/task_1061/org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/EqualsWithNullMultiQuickfixTest.xtend) is now complete pending review, but I'm still having trouble with [RemoveUnnecessaryModifiersMultiQuickfixTest](https://github.com/drkstr101/xtext-xtend/blob/task_1061/org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/RemoveUnnecessaryModifiersMultiQuickfixTest.xtend). I seem to be getting back less markers than I would be expecting.

Also FYI I had to fiddle with the `---` seperators a few times with multiple runs to get them right. I think once it even changed without any modification to the sample input. Maybe I did something to cause this without realizing, but a clean build seem to fix it. I will let you know if I spot any other quirks like that.

**Comment by miklossy** [comment]
@drkstr101 if you think that the EqualsWithNullMultiQuickfixTest is ready, please  provide a PR to easy the review.

**Comment by drkstr101** [comment]
@miklossy 
I was going to submit them both in a single PR once `RemoveUnnecessaryModifiersMultiQuickfixTest` was working. Would you prefer me to submit them separately? Either way is fine with me. My hope is that the PR will close this issue.

**Comment by miklossy** [comment]
@drkstr101: Creating one single PR that closes this issue is also fine for me.

Currently, the `MODEL_WITH_UNNECCESSARY_MODIFIERS` text also contains a not used variable. To make the `EqualsWithNullMultiQuickfixTest` succeed, you should modifiy the text that it contains only "Unnecessary modifiers warning" (they all can be fixed at once using multi-quickfix). You can do it e.g by adding the `println(A)` statement to the run method.

I can give you more feedback after the PR (containing both `EqualsWithNullMultiQuickfixTest` and `RemoveUnnecessaryModifiersMultiQuickfixTest`  test cases) has been created. Unfortunately, I cannot comment your code in your fork, only in a PullRequest. Moreover, the PR (before merge) will be built by the build system and the tests will be executed as well, so it will be also a valuable feedback whether the build server can pick up and execute the new test cases properly.

**Comment by drkstr101** [comment]
Thank you, I see what you mean. I opened up https://github.com/eclipse/xtext-xtend/pull/1084 and flagged it as *WIP*. I was hoping to have it wrapped up by noon (GMT-8) today.

**Comment by drkstr101** [comment]
@miklossy I ended up having to re-fork but everything looks ok now https://github.com/eclipse/xtext-xtend/pull/1085

**Comment by miklossy** [comment]
@drkstr101 : Thank you very much for your contribution!

Fixed in 2.23
